### ISSUE TITLE
Allow Renovate's azure-pipeline's manager to read azure repositories defined in resource blocks

### ISSUE THREAD

**Issue by JackRhodes** [renovatebot/renovate#15028]
### What would you like Renovate to be able to do?

Renovate has support for azure-pipelines, it does this by reading the [Resources block](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema/resources?view=azure-pipelines) defined in an azure pipelines yaml file

From this, it is able to update:

* Containers
* Git repositories

A git repository in azure pipeline has a `type` depending on where the repo is hosted, [docs here](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema/resources-repositories-repository?view=azure-pipelines): 

* `git`, (azure repository reference)
*  `github`,
*  `githubenterprise`,
*  bitbucket`.

Within Renovate's [current manager implementation](https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/azure-pipelines/extract.ts#L11) , it only supports `github` and will return null for all other types. This is not currently documented within Renovate and it is not clear that this is the case unless you view the implementation.

I would like to propose that support for `git` type is added, which will enable it to work within Azure Devops. Alongside this work, I would like to propose that the documentation is updated to state which repository types renovate supports in the manager to better define what it suppots.

GIVEN an azure pipeline, WHEN it has an azure pipeline repository reference in its resources THEN it should check if the repository has a newer version

Please let me know if any more detail is needed, or if I am misunderstanding anything.

### If you have any ideas on how this should be implemented, please tell us here.

(I am not a typescript dev and this is my first attempt at a suggestion)

If the repository reference is of type `git`, and the renovate platform is set to `azure` and has an `endpoint` defined, then we can construct the repository url reference.
This can be added to `lib/modules/manager/azure-pipelines/extract.ts`

I am unsure the cleanest way to achieve this as your would be mixing the manager domain with the platform domain, but I believe the actual change should be relatively simple code change.  Azure repositories can only be used within ADO so coupling the platform and manager is not as bad as it sounds.

I am happy to implement this, but would require guidance about the best way to handle this. 

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by rarkins** [comment]
Is it possible to create a public example on Azure DevOps?

**Comment by JackRhodes** [comment]
Yes I believe so. I'll try to setup a test ADO project with an example

**Comment by JackRhodes** [comment]
Apologies for the delay! I had to setup an ADO project which took a while

Renovate:

* [Example Renovate run](https://dev.azure.com/combarcapes/Renovate-Playground/_build/results?buildId=16&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=9c939e41-62c2-5605-5e05-fc3554afc9f5)
* [Renovate server setup](https://dev.azure.com/combarcapes/Renovate-Playground/_git/renovate-azure-pipelines?path=/azure-pipelines/renovate-pipeline.yaml)

The issue:

* [Affected yaml file](https://dev.azure.com/combarcapes/Renovate-Playground/_git/issue-15028?path=/azure-pipelines/issue.yaml) : this has an example of a github reference (that works), and an ADO git reference (that renovate does not generate for)
* [Pull Requests](https://dev.azure.com/combarcapes/Renovate-Playground/_git/issue-15028/pullrequests?_a=mine) : it has only generated a PR for the github dependency

Please let me know if anything is unclear!

**Comment by rarkins** [comment]
What is the expected result for the git dependency?

**Comment by JackRhodes** [comment]
In the [example](https://dev.azure.com/combarcapes/Renovate-Playground/_git/issue-15028?path=/azure-pipelines/issue.yaml&version=GBmain&line=4&lineEnd=5&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents) ,  renovate should be upping the version of `Renovate-Playground/example-repository` from `0.1.0` to `0.2.0`

**Comment by rarkins** [comment]
So it's defined as a "git" dependency yet it's implicitly pointing to Azure DevOps Services?

**Comment by JackRhodes** [comment]
Correct

[Docs here](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema/resources-repositories-repository?view=azure-pipelines#types)

**Comment by rarkins** [comment]
If the `git` reference is e.g. `MyBitbucket/vscode` then doesn't that mean we know the full name of the repo without needing to know anything about the current platform, endpoint or repository? i.e. it's always going to point to azure devops services?

**Comment by hypervtechnics** [comment]
as you can see in the docs it only will fetch from bitbucket if `type: bitbucket`. The type indicates rhe source of the repo. If `type: git` it will point to Azure DevOps (git sounds generic although Azure DevOps is meant)

**Comment by rarkins** [comment]
That doesn't answer my question

**Comment by hypervtechnics** [comment]
When you use `type: git` and `name: my-repository` it will point to `https://dev.azure.com/my-organization/my-project/_git/my-repository`. We can not infer my-project and my-organization from the azure pipelines file. That's only doable when we have the properties webUrl or remoteUrl available from the repository details.

**Comment by JackRhodes** [comment]
> If the `git` reference is e.g. `MyBitbucket/vscode` then doesn't that mean we know the full name of the repo without needing to know anything about the current platform, endpoint or repository? i.e. it's always going to point to azure devops services?

So yes, `git` will always point to the current azure devops organization. We know the full name of the repository, and potentially the project, but we do not have any way to see the organization from just the yaml file. Azure pipelines with a `git` reference will implicitly use the organization url for you since they sit in the same organization so it never gets explicitly defined in a pipeline

**Comment by rarkins** [comment]
In that case you can follow the example from GitLab: https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/gitlabci-include/extract.ts#L68

**Comment by hypervtechnics** [comment]
Thanks for the hint 😄 Unfortunetaly this works only in a subset of cases. We also have to know about the project name which is often omitted in the pipeline file if a repository in the same project is referenced. We would need to have the repository details available.

**Comment by JackRhodes** [comment]
@hypervtechnics For a first iteration, it could be ok to make it so you explicitly add the project name in the resource reference and not support implicit project lookups.

But yes, you are correct that for full behavior, then we would also need a way to lookup the current project that a repository is located

**Comment by hypervtechnics** [comment]
But I think we can already start discussion on handling it properly in the second iteration. Adding the project name everywhere is kind of a very huge task when you have a lot of projects and pipelines.

The only way I can think of is having the repository details (including the git clone uri) to fetch the project somewhat reliably. The endpoint url in the configuration is not suitable because it only points to the organization and not the project in the organization.

**Comment by HonkingGoose** [comment]
When you've created this feature, please edit the https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/azure-pipelines/readme.md file, to remove the part of the admonition that mentions feature 15028. 😉

**Comment by sheldonhull** [comment]
I'm a little confused on the state here. I reviewed the !20380 and indicates that as long as the `Project/Repo` format is followed (instead of also handling implicit project determination) then it should be merged in and working right?

I am running a test case with `"renovateVersion": "35.146.1"` docker version locally, and while one pipeline seems to be matched, it never takes any action on the out of date tag.
The docs also still reference this admonition. If i need to open another discussion instead I am willing, just thought this would be the appropriate follow-up place.

**Comment by 98christianp** [comment]
@sheldonhull you got it working as in the feature works for you? Could you share a working example, as I can't seem to get it working. It correctly gets the dependencies and current tag, but fails to see updates:
```
DEBUG: packageFiles with updates (repository=myorg/myrepo, baseBranch=main)
       "config": {
         "azure-pipelines": [
           {
             "deps": [
               {
                 "autoReplaceStringTemplate": "refs/tags/{{newValue}}",
                 "currentValue": "1.7.0",
                 "datasource": "git-tags",
                 "depName": "myproject/mytemplates",
                 "depType": "gitTags",
                 "packageName": "https://dev.azure.com/myorg/myproject/_git/myrepo",
                 "replaceString": "refs/tags/1.7.0",
                 "updates": [],
                 "versioning": "semver-coerced",
                 "warnings": [],
                 "sourceUrl": "https://dev.azure.com/myorg/myproject/_git/myrepo",
                 "currentVersion": "1.7.0",
                 "fixedVersion": "1.7.0"
               },
               ```

```
running a git tag -l
```bash
...

1.7.0
1.7.1
1.7.2
1.7.3
1.7.4
1.7.5
```
